### PR TITLE
Cqm parsers integration

### PR DESCRIPTION
**Pull request by daco101**
This PR uses cqm-parsers for measure loading and switched to the new measure model.
It also unifies the api vs api_v1 measure loading code.
It also adds some fixture loading and creation stuff.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] JIRA ticket for this PR: https://jira.mitre.org/browse/BONNIE-1915
- [x] JIRA ticket links to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] If UI changes have been made, google WAVE plug-in has been executed to ensure no 508 issues were introduced.
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Test fixtures updated and documented as necessary ( see [internal wiki](https://gitlab.mitre.org/bonnie/internal-documentation/wikis/testing#test-fixtures) )
- [x] Code coverage has not gone down and all code touched or added is covered. 
     * In rare situations, this may not be possible or applicable to a PR. In those situations:
         1. Note why this could not be done or is not applicable here: The rake tasks need to be rebuilt to use the new model
         2. Add TODOs in the code noting that it requires a test
         3. Add a JIRA task to add the test and link it here: https://jira.mitre.org/browse/BONNIE-1888
- [x] Automated regression test(s) pass N/A

**Reviewer 1:**

Name: @losborne 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

**Reviewer 2:**

Name: @hackrm 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

### CODE REVIEW

File: .rubocop_todo.yml [hunk 1]
```diff
@@ -1035,13 +1029,7 @@ Style/RedundantParentheses:
...
 # Configuration parameters: AllowMultipleReturnValues.
 Style/RedundantReturn:
-  Exclude:
-    - 'app/controllers/registrations_controller.rb'
-    - 'lib/measure_exported_results.rb'
-    - 'lib/patient_export.rb'
-    - 'lib/tasks/bonnie.rake'
-    - 'lib/version.rb'
-    - 'lib/util/fixture_exporter.rb'
+  Enabled: false
```
Comment (lukoktonos) [inline .rubocop_todo.yml:1032]: <3

File: app/controllers/api_v1/measures_controller.rb [hunk 2]
```diff
@@ -115,7 +117,7 @@ def show
       hash = {}
       http_status = 200
       begin
-        @api_v1_measure = CqlMeasure.by_user(current_resource_owner).where({:hqmf_set_id=> params[:id]}).sort_by { :updated_at }.first
+        @api_v1_measure = CQM::Measure.by_user(current_resource_owner).where({:hqmf_set_id=> params[:id]}).min_by { |m| m[:updated_at] }
```
Comment (lukoktonos) [inline app/controllers/api_v1/measures_controller.rb:120]: Nice!

File: app/controllers/api_v1/measures_controller.rb [hunk 3]
```diff
@@ -223,17 +232,26 @@ def create
...
+    rescue StandardError => e
+      render turn_exception_into_shared_error_if_needed(e).back_end_version
     end
 
     private
 
+    def retrieve_measure_details(params)
+      return {
+        'episode_of_care' => params[:calculation_type] == 'episode',
+        'calculate_sdes' => params[:calculate_sdes].nil? ? false : params[:calculate_sdes].to_s == 'true',
```
Comment (lukoktonos) [inline app/controllers/api_v1/measures_controller.rb]: I think `params[:calculate_sdes].to_s == 'true'` is enough--it should return false if `params[:calculate_sdes]` is `nil`:
`nil.to_s` -> `''`, `'' == 'true'` -> `false`

File: app/controllers/measures_controller.rb [hunk 4]
```diff
@@ -58,139 +42,45 @@ def create
...
-        # Throw error since component can't be deleted individually
-        render status: :bad_request, json: {error: "Component measures can't be deleted individually."}
-        return
-      elsif measure.composite
-        # If the measure if a composite, delete all the associated components
-        measure.component_hqmf_set_ids.each do |component_hqmf_set_id|
-          CqlMeasure.by_user(current_user).where(hqmf_set_id: component_hqmf_set_id).destroy
-        end
+    if measure.component
+      # Throw error since component can't be deleted individually
```
Comment (lukoktonos) [inline app/controllers/measures_controller.rb]: Comment is redundant with error message below and can be removed imo

File: app/controllers/populations_controller.rb [hunk 5]
```diff
@@ -1,13 +1,13 @@
 class PopulationsController < ApplicationController
 
   def update
-    # Added CQL Based Measure support for editing Population titles.
-    # If Measure doesn't exist in Measures move on to CqlMeasures.
-    measure = CqlMeasure.by_user(current_user).find(params[:measure_id])
-    measure.populations.at(params[:id].to_i)['title'] = params[:title]
+    measure = CQM::Measure.by_user(current_user).find(params[:measure_id])
+
+    pop_or_strat = measure.population_sets.select { |ps| ps.population_set_id == params[:id] }.first || measure.all_stratifications.select { |strat| strat.stratification_id == params[:id] }.first
```
Comment (lukoktonos) [inline app/controllers/populations_controller.rb:6]: Why the `.first`s on this line? Wouldn't that fail to update ones that aren't first?

File: test/vcr_setup.rb [hunk 6]
```diff
@@ -17,4 +17,18 @@
...
 
+  # Add a custom matcher for use with the bulk request by typheous, so we can ignore service ticket
+  c.register_request_matcher :uri_no_st do |req1, req2|
+    remove_service_ticket_from_uri(req1.uri) == remove_service_ticket_from_uri(req2.uri)
+  end
+end
+
+def remove_service_ticket_from_uri(uri)
+  uri = String.new(uri)
+  st = uri[/(&ticket=.*)&|(&ticket=.*)$/,0]
```
Comment (lukoktonos) [inline test/vcr_setup.rb]: `st` -> `service_ticket` (st is too close to string imo)

File: app/helpers/measure_helper.rb [hunk 7]
```diff
@@ -117,6 +270,66 @@ def self.measures_population_update(measures, is_update, current_user, measure_d
         patient.update_expected_value_structure!(measure)
       end
     end
+  end
+
+  def add_measures_to_user(measures, current_user)
+    # what is this for?
```
Comment (lukoktonos) [inline app/helpers/measure_helper.rb]: remove comment

File: app/helpers/measure_helper.rb [hunk 8]
```diff
@@ -117,6 +270,66 @@ def self.measures_population_update(measures, is_update, current_user, measure_d
...
+    measures.each { |measure| current_user.cqm_measures << measure }
     current_user.save!
   end
+
+  def delete_for_update(existing, user)
+    # TODO: make sure this is good.
```
Comment (lukoktonos) [inline app/helpers/measure_helper.rb]: Is this still in investigation?

File: app/helpers/measure_helper.rb [hunk 9]
```diff
@@ -117,6 +270,66 @@ def self.measures_population_update(measures, is_update, current_user, measure_d
...
     end
+  end
+
+  def add_measures_to_user(measures, current_user)
+    # what is this for?
+    measures.each { |measure| current_user.cqm_measures << measure }
     current_user.save!
   end
+
+  def delete_for_update(existing, user)
```
Comment (lukoktonos) [inline app/helpers/measure_helper.rb:275]: `delete_for_update` -> `delete_component_measures_for_update`

File: app/helpers/measure_helper.rb [hunk 10]
```diff
@@ -117,6 +270,66 @@ def self.measures_population_update(measures, is_update, current_user, measure_d
...
+      errored_measure_file.write(uploaded_file.read)
+      uploaded_file.close
+    end
+    File.chmod(0o644, File.join(errors_dir, filename))
+    File.open(File.join(errors_dir, "#{clean_email}_#{Time.now.strftime('%Y-%m-%dT%H%M%S')}.error"), 'w') do |f|
+      f.write("Original Filename was #{measure_file.original_filename}\n")
+      f.write(error.to_s + "\n" + (error.backtrace||[]).join("\n"))
+    end
+    # email the error
+    if error.respond_to?(:operator_error) && error.operator_error && defined? ExceptionNotifier::Notifier # rubocop:disable Style/GuardClause
```
Comment (lukoktonos) [inline app/helpers/measure_helper.rb]: Is this rubocop comment still needed with your changes in the rubocop file above?

File: app/helpers/measure_helper.rb [hunk 11]
```diff
@@ -117,6 +270,66 @@ def self.measures_population_update(measures, is_update, current_user, measure_d
...
+      # params[:error_file] = filename  #TODO what is this
```
Comment (lukoktonos) [inline app/helpers/measure_helper.rb]: Is this TODO still under investigation?

**Review by lukoktonos — CHANGES_REQUESTED** [review]

File: app/helpers/measure_helper.rb (hunk 9 above)
Comment (daco101) [reply app/helpers/measure_helper.rb:275]: this method deletes the measure and any components

File: app/helpers/measure_helper.rb (hunk 7 above)
Comment (daco101) [reply app/helpers/measure_helper.rb]: do you know what this is for? why do we manually maintain a list of measures on a user?

File: app/controllers/populations_controller.rb (hunk 5 above)
Comment (daco101) [reply app/controllers/populations_controller.rb:6]: The ids should be unique, so we only expect to get one pop_or_strat.

File: app/helpers/measure_helper.rb (hunk 7 above)
Comment (lukoktonos) [reply app/helpers/measure_helper.rb]: It's the mongoid relationship between the user and the measures.  A User `has_many :cqm_measures`

Comment (daco101) [reply app/helpers/measure_helper.rb]: I've confirmed that we do not need to manually maintain this, mongoid will take care of it.

Comment (lukoktonos) [reply app/helpers/measure_helper.rb]: Interesting--so we can remove this then?

File: app/controllers/api_v1/measures_controller.rb [hunk 12]
```diff
@@ -223,17 +232,26 @@ def create
...
+      render turn_exception_into_shared_error_if_needed(e).back_end_version
     end
 
     private
 
+    def retrieve_measure_details(params)
+      return {
+        'episode_of_care' => params[:calculation_type] == 'episode',
+        'calculate_sdes' => params[:calculate_sdes].to_s == 'true',
+        'population_titles' => params[:population_titles]
```
Comment (hackrm) [inline app/controllers/api_v1/measures_controller.rb:251]: Is there a reason why 'population_titles' is included in this _retrieve_measure_details()_ function but not in [this one](https://github.com/projecttacoma/bonnie/pull/1179/files#diff-9f6c25d2ea5170e4953a469c855a47d4R113) from app/controllers/measures_controller.rb?

Comment (hossenlopp) [reply app/controllers/api_v1/measures_controller.rb:251]: The API allows the API user to optionally provide the titles for the populations on upload. In the Bonnie frontend the measure upload does not ask for them. We ask the user after it uploads if they want to change the titles from a default.

Comment (hackrm) [reply app/controllers/api_v1/measures_controller.rb:251]: Thanks for the explanation!

File: app/helpers/measure_helper.rb (hunk 10 above)
Comment (daco101) [reply app/helpers/measure_helper.rb]: Yes, im open to suggestions though.

File: app/helpers/measure_helper.rb (hunk 11 above)
Comment (daco101) [reply app/helpers/measure_helper.rb]: I just decided to delete the line. I can't see that it ever did anything.

**Review by lukoktonos — APPROVED** [review]
Looks good. Manual testing is showing consistent performance now with quite a significant speed up in measure upload.  Great work!

**Comment by hackrm** [comment]
I know some tests are skipped for work that will be done in the future. If this hasn't been done already, would it be helpful add references to these skipped test to the description of those relevant tasks? That way the person completing the task knows to go back and update those specific tests.

**Review by hackrm — APPROVED** [review]
Besides the little tip I left earlier, it looks good to me!

**Pull request merged**